src/Http/MiddlewareQueue.php

```diff
@@ -82,25 +85,32 @@ protected function resolve(int $index): ?MiddlewareInterface
             return null;
         }
 
-        if (is_string($this->queue[$index])) {
-            $class = $this->queue[$index];
-            $className = App::className($class, 'Middleware', 'Middleware');
+        $middleware = $this->queue[$index];
+        if (is_string($middleware)) {
+            $className = App::className($middleware, 'Middleware', 'Middleware');
             if ($className === null || !class_exists($className)) {
                 throw new RuntimeException(sprintf(
                     'Middleware "%s" was not found.',
-                    $class
+                    $middleware
                 ));
             }
-            $callable = new $className();
-        } else {
-            $callable = $this->queue[$index];
+            $middleware = new $className();
         }
 
-        if (!$callable instanceof MiddlewareInterface) {
-            $callable = new DoublePassMiddleware($callable);
+        if ($middleware instanceof MiddlewareInterface) {
+            return $this->middlewares[$index] = $middleware;
         }
 
-        return $this->middlewares[$index] = $callable;
+        if (!$middleware instanceof Closure) {
+            return $this->middlewares[$index] = new DoublePassDecoratorMiddleware($middleware);
+        }
+
+        $info = new ReflectionFunction($middleware);
+        if ($info->getNumberOfParameters() > 2) {
+            return $this->middlewares[$index] = new DoublePassDecoratorMiddleware($middleware);
+        }
```

Comment:
This is good. Makes adding middleware backwards compatible and easier to use in the future as well.